### PR TITLE
machines: Implement creation of VM snapshots

### PR DESCRIPTION
**Pull request by skobyda**
- [x] This PR builds on top of #13100 and can be merged only after it. Only last commit is relevant to this PR.

Screenshot from 2020-06-22 14-57-06

So far user can only input name of snapshot and description.
Name is pre-generated from VM name + date and time, and description is optional input, so ideally user doesn't have to input anything an can straight go to clicking "Create".
Screenshot from 2020-06-22 14-54-02

I also experimented here with how we evaluate  incorrect inputs in dialogs, and tried to us React's onBlur trigger. This means that whenever uses looses trigger of a input field (clicks outside of a input field or start typing into another input field), input field is evaluated. This is more immediate and more direct compared to evaluating inputs only after user click "Create".
Screenshot from 2020-06-22 14-54-23

### CODE REVIEW

**Comment by skobyda** [comment]
No failures apart from one which can be see in other unrelated PRs, to it's safe to assume it's not related to this PR.

File: pkg/machines/components/vmSnapshotsCreateModal.jsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    onValidate(subject) {
+        const { name, validationError } = this.state;
+        const { vm } = this.props;
+
+        if (subject === "name") {
+            const newValidationError = { ...validationError };
+            if (vm.snapshots.findIndex(snap => snap.name === name) > -1)
+                newValidationError.name = "Snapshot with this name already exists";
```
Comment (andreasn) [inline pkg/machines/components/vmSnapshotsCreateModal.jsx:107]: I would prefer the shorter _name already exists_.
Certain languages, such as German, has a tendency to get very long in translations.
It is also clear from the context that this is about snapshots.

File: pkg/machines/components/vmSnapshotsTab.css [hunk 2]
```diff
@@ -0,0 +1,12 @@
+.no-snapshots {
+    text-align: center;
+    line-height: 3.0rem
+}
+
+.snap-greyed-out {
+    color: #72767b;
```
Comment (andreasn) [inline pkg/machines/components/vmSnapshotsTab.css:7]: I think we might have a color variable for this. Maybe `--pf-global--palette--black-600`?
https://www.patternfly.org/v4/documentation/overview/global-css-variables#global-css-variables

File: pkg/machines/components/vmSnapshotsCreateModal.jsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+const DescriptionRow = ({ onValueChanged, dialogValues }) => {
+    return (
+        <>
+            <label className="control-label" htmlFor="description">
+                {_("Description")}
+            </label>
+            <TextArea value={dialogValues.description}
+                id="description"
+                placeholder={_("Optional arbitrary description of a snapshot")}
```
Comment (andreasn) [inline pkg/machines/components/vmSnapshotsCreateModal.jsx:68]: This is very complex wording. _Description_ is already mentioned in the label, and Snapshot is clear from the context.
I would suggest simply _Optional_

File: pkg/machines/components/vmSnapshotsCreateModal.jsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        <>
+            <label className="control-label" htmlFor="name">
+                {_("Name")}
+            </label>
+            <FormGroup validationState={dialogValues.validationError.name ? "error" : undefined} controlId="name">
+                <TextInput value={dialogValues.name}
+                    id="name"
+                    type="text"
+                    onChange={(value) => onValueChanged("name", value)}
+                    onBlur={() => onValidate("name")}
```
Comment (andreasn) [inline pkg/machines/components/vmSnapshotsCreateModal.jsx:47]: "I also experimented here with how we evaluate incorrect inputs in dialogs, and tried to us React's onBlur trigger. This means that whenever uses looses trigger of a input field (clicks outside of a input field or start typing into another input field), input field is evaluated. This is more immediate and more direct compared to evaluating inputs only after user click "Create"."

I chatted with Jenn a little bit about this. While the [Patternfly guidelines for error validation](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/forms#errors-and-validation) suggests both methods, without taking a strong stance on either, we tend to only use Error-on-submit in Cockpit. However, Composer uses error-on-focus, and it would be good to be consistent between the pages.
Could we skip this for now, until we get the chance to take a proper decision for all across Cockpit? I think it would be bad for the Machines page itself to be internally inconsistent.

File: pkg/machines/components/vmSnapshotsTab.jsx [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        this.setState({ showCreateSnapshotModal: false });
+    }
+
+    render() {
+        const { vm, dispatch } = this.props;
+        const id = vmId(vm.name);
+
+        const emptyCaption = (<div className="no-snapshots">
+            <h3>{_("No snapshots")}</h3>
+            {_("Snapshots allow you to revert to an earlier state if something goes wrong")}
```
Comment (andreasn) [inline pkg/machines/components/vmSnapshotsTab.jsx:79]: But only if you take them!
If this string is aimed towards people unfamiliar with snapshot functionality, I the the text need to be a little bit more specific. Can't think of a good exact string right now, though.

File: pkg/machines/components/vmSnapshotsTab.jsx [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+import { InfoAltIcon } from '@patternfly/react-icons';
+
+import './vmSnapshotsTab.css';
+
+const _ = cockpit.gettext;
+
+function prettyTime(unixTime) {
+    moment.locale(cockpit.language, {
+        longDateFormat : {
+            LT: "HH:mm:ss",
```
Comment (andreasn) [inline pkg/machines/components/vmSnapshotsTab.jsx:37]: (I don't know if this actually gets set here, but as a non-programmer, I'm kind of shit at reading code, I'll put this comment here :)

I don't think we need the seconds here.
I also have all other time/date occurrences on this test VM formatted as AM/PM, but the snapshots seems to stick to 24-hour clock.

File: pkg/machines/components/vmSnapshotsTab.jsx (hunk 5 above)
Comment (skobyda) [reply pkg/machines/components/vmSnapshotsTab.jsx:79]: @KKoukiou also had a comment about that line here: https://github.com/cockpit-project/cockpit/pull/13100#discussion_r444889834
Perhaps we can discuss it there.

File: pkg/machines/components/vmSnapshotsCreateModal.jsx (hunk 4 above)
Comment (skobyda) [reply pkg/machines/components/vmSnapshotsCreateModal.jsx:47]: Alright. I was thinking about rewriting it everywhere else, so I was think about experimenting with it here.
But we can have separate PR where we do it all at once.

File: pkg/machines/components/vmSnapshotsTab.jsx (hunk 6 above)
Comment (skobyda) [reply pkg/machines/components/vmSnapshotsTab.jsx:37]: Fixed this and all above mentioned

**Comment by skobyda** [comment]
@KKoukiou could you also give review to this? Only last commit is relevant to this PR.

**Comment by martinpitt** [comment]
#13100 landed; so this can be rebased and finished.

**Comment by skobyda** [comment]
Rebased. @andreasn and @KKoukiou already gave approval to this PR, so since I done no changes and only rebased I think this can land as soon as tests pass.

File: pkg/machines/components/vmSnapshotsCreateModal.jsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+import cockpit from "cockpit";
+import React from "react";
+import moment from "moment";
+
+import { Modal, FormGroup, HelpBlock } from "patternfly-react";
```
Comment (KKoukiou) [inline pkg/machines/components/vmSnapshotsCreateModal.jsx:23]: I know we use it elsewhere, but for future PRs you can start using Forms and HelpText from PF4. :)

**Comment by garrett** [comment]
I still cannot see VM snapshots even when I have this branch compiled.

It looks like this (there's no tab for snapshots):
Screenshot_2020-07-23 Virtual Machines - garrett Rain

**Comment by KKoukiou** [comment]
> I still cannot see VM snapshots even when I have this branch compiled.
> 
> It looks like this (there's no tab for snapshots):
> Screenshot_2020-07-23 Virtual Machines - garrett Rain

You can only test it in debian-testing VM for now because of the libvirt-dbus version.

**Comment by skobyda** [comment]
> I still cannot see VM snapshots even when I have this branch compiled.
> 
> It looks like this (there's no tab for snapshots):
> Screenshot_2020-07-23 Virtual Machines - garrett Rain

I guess you are using fedora, so now you still ave libvirt-dbus 1.3.0 Snapshots APIs are in 1.4.0
Looking at https://repology.org/project/libvirt-dbus/versions it seems new libvirt-dbus version got it way into Fedora Rawhide too.
You can run upstream version of libvirt-dbus. Clone repo from https://github.com/libvirt/libvirt-dbus/ and then inside repo run:
```
meson build
ninja -C build
cd build
./run ./src/libvirt-dbus
```

**Comment by skobyda** [comment]
> Is "arbitrary" still in the textarea? (It shouldn't be.)

Not i just says "Optional"

**Comment by garrett** [comment]
Additionally, the placeholder text should only ever be used as an example: https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/forms#placeholder-text

PatternFly generally calls out mandatory fields, not optional ones... but Cockpit usually doesn't label mandatory fields. So I guess if we show that the description field is optional, we could do that with helper text below it? Or start listing mandatory fields?

Also, is there an `<hr>` between the name and description? There shouldn't be. They're both describing the snapshot and are related to each other. (It looks like there's some extra space between the form elements.)

File: pkg/machines/components/vmSnapshotsCreateModal.jsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        }
+    }
+
+    render() {
+        const { idPrefix, onClose } = this.props;
+
+        const body = (
+            <form className="ct-form">
+                <NameRow dialogValues={this.state} onValueChanged={this.onValueChanged} />
+
```
Comment (KKoukiou) [inline pkg/machines/components/vmSnapshotsCreateModal.jsx:134]: Do you really want these new lines here?

File: pkg/machines/components/vmSnapshotsCreateModal.jsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                    type="text"
+                    onChange={(value) => onValueChanged("name", value)}
+                    aria-label={_("Name input text")}
+                />
+                { dialogValues.validationError.name &&
+                <HelpBlock>
+                    <p className="text-danger">{dialogValues.validationError.name}</p>
+                </HelpBlock> }
+            </FormGroup>
+
```
Comment (KKoukiou) [inline pkg/machines/components/vmSnapshotsCreateModal.jsx:54]: Please remote this new line.

File: pkg/machines/components/vmSnapshotsTab.jsx [hunk 10]
```diff
@@ -43,118 +44,153 @@ function prettyTime(unixTime) {
...
+
+                    return (
+                        <div id={`${id}-snapshot-${snapId}-description`}>
+                            {desc}
+                        </div>
+                    );
+                }
+            },
+            {
+                name: _("State"), value: (snap, snapId) => {
```
Comment (KKoukiou) [inline pkg/machines/components/vmSnapshotsTab.jsx:109]: Isn't this change unrelated?

File: pkg/machines/components/vmSnapshotsTab.jsx [hunk 11]
```diff
@@ -43,118 +44,153 @@ function prettyTime(unixTime) {
...
+                    return (
+                        <div id={`${id}-snapshot-${snapId}-type`}>
+                            {state}
+                            {tooltip}
+                        </div>
+                    );
+                }
+            },
+            {
+                name: _("Parent snapshot"), value: (snap, snapId) => {
```
Comment (KKoukiou) [inline pkg/machines/components/vmSnapshotsTab.jsx:139]: These  capitalization changes should be in different commit.

File: pkg/machines/components/vmSnapshotsTab.jsx (hunk 10 above)
Comment (skobyda) [reply pkg/machines/components/vmSnapshotsTab.jsx:109]: oh the changes i just done in previous PR gets overwritten. Thanks for catching that!

File: pkg/machines/components/vmSnapshotsTab.jsx (hunk 11 above)
Comment (skobyda) [reply pkg/machines/components/vmSnapshotsTab.jsx:139]: Fixed this and all above

File: pkg/machines/components/vmSnapshotsCreateModal.jsx (hunk 7 above)
Comment (skobyda) [reply pkg/machines/components/vmSnapshotsCreateModal.jsx:23]: I updated it to PF4

**Comment by skobyda** [comment]
> Additionally, the placeholder text should only ever be used as an example: https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/forms#placeholder-text
> 
> PatternFly generally calls out mandatory fields, not optional ones... but Cockpit usually doesn't label mandatory fields. So I guess if we show that the description field is optional, we could do that with helper text below it? Or start listing mandatory fields?
>
> Also, is there an `<hr>` between the name and description? There shouldn't be. They're both describing the snapshot and are related to each other. (It looks like there's some extra space between the form elements.)

So I put helper text below (marking all mandatory field seems like a big change which should be done globally and got rid of the <hr>:
Screenshot from 2020-07-23 13-50-43

**Comment by garrett** [comment]
The help text isn't correct. I can tell because the spacing and the text size is not correct. That said, having text hovering down there looks quite odd.

Is it OK to drop the optional text right now and get the PR in as-is?

File: pkg/machines/components/vmSnapshotsCreateModal.jsx (hunk 8 above)
Comment (KKoukiou) [reply pkg/machines/components/vmSnapshotsCreateModal.jsx:134]: You forgot to remove the newlines ;)

File: pkg/machines/components/vmSnapshotsTab.jsx [hunk 12]
```diff
@@ -43,118 +44,153 @@ function prettyTime(unixTime) {
...
-        rows = vm.snapshots.sort((a, b) => b.creationTime - a.creationTime).map((target, snapId) => {
-            const columns = detailMap.map(d => {
-                let column = null;
-                if (typeof d.value === 'string') {
-                    if (target[d.value] !== undefined)
-                        column = { title: <div id={`${id}-snapshot-${snapId}-${d.value}`}>{target[d.value]}</div> };
-                }
-                if (typeof d.value === 'function')
-                    column = { title: d.value(target, snapId) };
+class VmSnapshotsTab extends React.Component {
```
Comment (KKoukiou) [inline pkg/machines/components/vmSnapshotsTab.jsx:47]: How about using React Hooks (useState) and keeping this component functional?

This will make the diff nice mainly.

Comment (skobyda) [reply pkg/machines/components/vmSnapshotsTab.jsx:47]: The another PR https://github.com/cockpit-project/cockpit/pull/14283 also adds stateful logic into this class, so perhaps it makes sense to turn it into class once and for all.
But thanks, for the future, I will keep in my to not change functional components into classes because of few changes.

**Comment by skobyda** [comment]
@garrett I got rid of the "Optional" helper text. BTW what's wrong with that text? I'm I using it incorrectly or we have no good way to use it in cockpit yet (like we are missing some css for that)? If so, it would be good to create an issue for that.

**Comment by KKoukiou** [comment]
For release notes: (a hint - ask an en native speaker for improvement)
Screen Shot 2020-07-27 at 11 44 09

### Machines: Creating VM snapshots

Libvirt supports two types of snapshots, [internal and external](https://libvirt.org/formatsnapshot.html), but external are not yet in feature parity with the internal. For this reason, we decided to go with the internal snapshots in Cockpit UI. Because of using internal snapshots this feature only works with qcow2 disk images. VM snapshots are supported for running VMs, in which case the snapshot will combine disk and VM memory state, as well as  offline VMs, in which case only the disk contents will be saved.

**Comment by KKoukiou** [comment]
@skobyda seems the new tests is flaky? 
https://logs.cockpit-project.org/logs/pull-14259-20200727-091651-8ffe424f-debian-testing/log.html#301-2

I will retry for comparison.

**Comment by KKoukiou** [comment]
@skobyda the new test still needs adjustment - sorry :( 
https://logs.cockpit-project.org/logs/pull-14259-20200727-131820-dcd3e42e-debian-testing/log.html#301-2
